x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFeatures.java

```diff
@@ -0,0 +1,229 @@
+        this.incrementalSyncEnabled = incrementalSyncEnabled;
+        this.syncRulesAdvancedEnabled = syncRulesAdvancedEnabled;
+        this.syncRulesBasicEnabled = syncRulesBasicEnabled;
+    }
+
+    public ConnectorFeatures(StreamInput in) throws IOException {
+        this.documentLevelSecurityEnabled = in.readOptionalBoolean();
+        this.filteringAdvancedConfigEnabled = in.readOptionalBoolean();
+        this.filteringRulesEnabled = in.readOptionalBoolean();
+        this.incrementalSyncEnabled = in.readOptionalBoolean();
+        this.syncRulesAdvancedEnabled = in.readOptionalBoolean();
+        this.syncRulesBasicEnabled = in.readOptionalBoolean();
+    }
+
+    public static final ParseField DOCUMENT_LEVEL_SECURITY_ENABLED_FIELD = new ParseField("document_level_security_enabled");
+    public static final ParseField FILTERING_ADVANCED_CONFIG_ENABLED_FIELD = new ParseField("filtering_advanced_config_enabled");
+    public static final ParseField FILTERING_RULES_ENABLED_FIELD = new ParseField("filtering_rules_enabled");
+    public static final ParseField INCREMENTAL_SYNC_ENABLED_FIELD = new ParseField("incremental_sync_enabled");
+    public static final ParseField SYNC_RULES_ADVANCED_ENABLED_FIELD = new ParseField("sync_rules_advanced_enabled");
+    public static final ParseField SYNC_RULES_BASIC_ENABLED_FIELD = new ParseField("sync_rules_basic_enabled");
+
+    private static final ConstructingObjectParser<ConnectorFeatures, Void> PARSER = new ConstructingObjectParser<>(
+        "connector_features",
+        true,
+        args -> new Builder().setDocumentLevelSecurityEnabled((Boolean) args[0])
+            .setFilteringAdvancedConfig((Boolean) args[1])
+            .setFilteringRules((Boolean) args[2])
+            .setIncrementalSyncEnabled((Boolean) args[3])
+            .setSyncRulesAdvancedEnabled((Boolean) args[4])
+            .setSyncRulesBasicEnabled((Boolean) args[5])
+            .build()
+    );
+
+    static {
+        PARSER.declareBoolean(optionalConstructorArg(), DOCUMENT_LEVEL_SECURITY_ENABLED_FIELD);
+        PARSER.declareBoolean(optionalConstructorArg(), FILTERING_ADVANCED_CONFIG_ENABLED_FIELD);
+        PARSER.declareBoolean(optionalConstructorArg(), FILTERING_RULES_ENABLED_FIELD);
+        PARSER.declareBoolean(optionalConstructorArg(), INCREMENTAL_SYNC_ENABLED_FIELD);
+        PARSER.declareBoolean(optionalConstructorArg(), SYNC_RULES_ADVANCED_ENABLED_FIELD);
+        PARSER.declareBoolean(optionalConstructorArg(), SYNC_RULES_BASIC_ENABLED_FIELD);
+    }
+
+    public static ConnectorFeatures fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        {
+            if (documentLevelSecurityEnabled != null) {
+                builder.startObject("document_level_security");
```

Comment:
we can reuse DOCUMENT_LEVEL_SECURITY_ENABLED_FIELD.getPreferredName() here to be consistent

Reply:
I re-used names where I could, but the nested nature of some older feature fields (lovely tech debt ;) ) makes it required to use different names. 

This parser will be use for parsing json body coming from an API call in future PRs.